### ISSUE TITLE
SentryHttpMessageHandler not being added when AddHttpClient is before UseSentry

### ISSUE THREAD

**Issue by mattjohnsonpint** [getsentry/sentry-dotnet#2373]
### Package

Sentry.AspNetCore

### .NET Flavor

.NET

### .NET Version

7.0.5

### OS

Any (not platform specific)

### SDK Version

3.31.0

### Steps to Reproduce

An an ASP.NET Core app, such as `Sentry.Samples.AspNetCore.Basic` use `IHttpClientFactory` per [ms docs](https://learn.microsoft.com/aspnet/core/fundamentals/http-requests) to make an outbound request from within a web controller.

### Expected Result

The `SentryHttpMessageHandler` should automatically be connected by use of the `UseSentry` call from `Sentry.AspNetCore`.  The order that it is added in relation to `AddHttpClient` should not matter.

### Actual Result

It only works if `AddHttpClient` is after `UseSentry`.  If it's _before_ `UseSentry`, then the message handler is not added, and HTTP Client spans are not auto-instrumented.

This works.  `HttpClient` requests will be auto-instrumented in Sentry.

```csharp
builder.WebHost.UseSentry();
builder.Services.AddHttpClient();
```

This does not, but should.  `HttpClient` requests succeed, but are not auto-instremented in Sentry.
```csharp
builder.Services.AddHttpClient();
builder.WebHost.UseSentry();
```

## Further info

`SentryHttpMessageHandler` gets wired up for outgoing requests in `SentryWebHostBuilderExtensions`. Specifically, this is done when configuring logging via:

```
var sentryBuilder = logging.Services.AddSentry();
```

Eventually that calls `Sentry.Extensions.Logging.ServiceCollectionExtensions.AddSentry` which contains this call that refers to https://github.com/getsentry/sentry-dotnet/issues/785:

```
// Custom handler for HttpClientFactory.
// Must be singleton: https://github.com/getsentry/sentry-dotnet/issues/785
services.TryAddSingleton<IHttpMessageHandlerBuilderFilter, SentryHttpMessageHandlerBuilderFilter>();
```

**Comment by mattjohnsonpint** [comment]
Middleware order matters,  but service registration order should not.


I don't think any ASP.NET middleware is involved here.  Its just registering `SentryHttpMessageHandlerBuilderFilter` with the service provider.

**Comment by jamescrosswell** [comment]
Created a minimal ASP.NET Core webapp with a custom `DelegatingHandler` that gets registered via a custom `IHttpMessageHandlerBuilderFilter` and with that app the order of service registration has no effect (the custom message handler gets invoked either way). 

There must be something funky about how we're wiring things up then.

**Comment by jamescrosswell** [comment]
On a whim, I tried changing this line in our code:

```
        services.TryAddSingleton<IHttpMessageHandlerBuilderFilter, SentryHttpMessageHandlerBuilderFilter>();
```

To this instead...

```
        services.AddSingleton<IHttpMessageHandlerBuilderFilter, SentryHttpMessageHandlerBuilderFilter>();
```

And it "fixes" the issue. It's not what I was expecting. Without tracing into what the `TryAddSingleton` method is doing (and why it's not registering our `SentryHttpMessageHandlerBuilderFilter` when the `AddSingleton` method is), it's hard to puzzle this one out.

What are the dangers of using AddSingleton here vs TryAddSingleton? If it's about having our filter run multiple times on outbound requests, we could potentially mitigate with some code in the `SentryHttpMessageHandlerBuilderFilter`. For example (haven't tried this... just hypothetically):

```
    public Action<HttpMessageHandlerBuilder> Configure(Action<HttpMessageHandlerBuilder> next) =>
        handlerBuilder =>
        {
            var hub = _getHub();
            if (!handlerBuilder.AdditionalHandlers.Any(h => h is SentryHttpMessageHandler))
            {
                handlerBuilder.AdditionalHandlers.Add(
                    new SentryHttpMessageHandler(hub)
                );
            }

            next(handlerBuilder);
        };
```

**Comment by mattjohnsonpint** [comment]
That sounds like a good approach.  We should just add the filter (not try-add), and the filter builder can ensure the message handler is only added once.

Nice find!

**Comment by fbeauche** [comment]
Any ETA when this fix will be included in a release ? ( i guess it will be in 3.34 ? )

**Comment by mattjohnsonpint** [comment]
@fbeauche  - Yes, we'll be having a 3.34.0 release later this week that will include this fix and the others that are staged (see the "unreleased" section of the [changelog](https://github.com/getsentry/sentry-dotnet/blob/main/CHANGELOG.md)), along with a couple of items that are in-progress and expected to be completed soon.

In the meantime though, you can avoid the original bug by making sure that `AddHttpClient` comes after `UseSentry` in your code.

**Comment by mattjohnsonpint** [comment]
Small change of plans. We decided to release this and the other staged fixes in 3.33.1, which is available now. Thanks.
